exist-core/src/main/java/org/exist/xquery/functions/util/CollectionSuspendTriggers.java

```diff
@@ -0,0 +1,126 @@
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+ *
+ *  $Id$
+ */
+package org.exist.xquery.functions.util;
+
+import java.net.URISyntaxException;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.exist.collections.CollectionConfiguration;
+import org.exist.collections.LockedCollection;
+import org.exist.dom.persistent.NodeProxy;
+import org.exist.dom.QName;
+import org.exist.security.PermissionDeniedException;
+import org.exist.storage.DBBroker;
+import org.exist.xmldb.XmldbURI;
+import org.exist.xquery.BasicFunction;
+import org.exist.xquery.Cardinality;
+import org.exist.xquery.FunctionSignature;
+import org.exist.xquery.XPathException;
+import org.exist.xquery.XQueryContext;
+import org.exist.xquery.value.*;
+import org.xmldb.api.base.Collection;
+import org.xmldb.api.base.XMLDBException;
+
+/**
+ * @author <a href="mailto:wolfgang@exist-db.org">Wolfgang Meier</a>
+ *
+ */
+public class CollectionSuspendTriggers extends BasicFunction {
+
+    protected static final Logger logger = LogManager.getLogger(CollectionName.class);
+    private final static QName QN_COLLECTION_SUSPEND_TRIGGERS = new QName("collection-suspend-triggers", UtilModule.NAMESPACE_URI, UtilModule.PREFIX);
+    private final static QName QN_COLLECTION_ENABLE_TRIGGERS = new QName("collection-enable-triggers", UtilModule.NAMESPACE_URI, UtilModule.PREFIX);
+
+    public final static FunctionSignature signatures[] = {
+            new FunctionSignature(
+                    QN_COLLECTION_SUSPEND_TRIGGERS,
+                    "suspends triggers for a collection.",
+                    new SequenceType[]{
+                            new FunctionParameterSequenceType("node-or-path-string", Type.ITEM, Cardinality.ZERO_OR_ONE, "The document node or a path string.")
+                    },
+                    new FunctionReturnSequenceType(Type.BOOLEAN, Cardinality.ONE, "current state of triggers for the collection")
+
+            ),
+            new FunctionSignature(
+                    QN_COLLECTION_ENABLE_TRIGGERS,
+                    "re-enables possibly suspended triggers for a collection.",
+                    new SequenceType[]{
+                            new FunctionParameterSequenceType("node-or-path-string", Type.ITEM, Cardinality.ZERO_OR_ONE, "The document node or a path string.")
+                    },
+                    new FunctionReturnSequenceType(Type.BOOLEAN, Cardinality.ONE, "current state of triggers for the collection")
+
+            )
+    };
+
+
+    public CollectionSuspendTriggers(XQueryContext context, FunctionSignature signature) {
+        super(context, signature);
+    }
+
+    /* (non-Javadoc)
+     * @see org.exist.xquery.BasicFunction#eval(org.exist.xquery.value.Sequence[], org.exist.xquery.value.Sequence)
+     */
+    public Sequence eval(Sequence[] args, Sequence contextSequence)
```

Comment:
IMHO there are are a lot of bad things happening in this function:

1. This can be used by any user to disable all triggers for all brokers. I could issue a series of REST calls which would break anyones app!
2. This disables all triggers for a collection, and is not per-broker. 
3. Disabling triggers should be a temporary measure, as such the design of this function leads to problems with a disable-enable pattern. I would suggest it instead should be refactored as a HoF.

Reply:
@adamretter the code ypu are referring to was moved to #3339, let's continue discussion there.
- 1 - can you sketch out that attack scenario? while testing I saw `PermissionDenied` exceptions, so I assumed standard eXist-db access control applies;
- 2 - Yes. That is the purpose of #3339 
- 3 - Fully agree. In a perfect world we would have LISP-style `(with-triggers-disabled (code ...))` where the `with-...` clause does all the setup and teardown actions around `code`. I fear this may not work with multithreaded eXist-db and asynchrnous HTTP I/O, but I'm open for suggestions